File: backend/src/utils/security.ts
import {
  createCipheriv,
  createDecipheriv,
  createHash,
  createHmac,
  randomBytes
} from "node:crypto";
import bcrypt from "bcryptjs";
import { SignJWT, jwtVerify } from "jose";
import { config, getFernetKeyBytes } from "../config";

const encoder = new TextEncoder();
const keyMaterial = getFernetKeyBytes();
const signingKey = keyMaterial.subarray(0, 16);
const encryptionKey = keyMaterial.subarray(16, 32);
const jwtKey = encoder.encode(config.appSecret);

function fernetTimestamp(date = new Date()): Buffer {
  const timestamp = Math.floor(date.getTime() / 1000);
  const bytes = Buffer.alloc(8);
  bytes.writeBigUInt64BE(BigInt(timestamp));
  return bytes;
}

export function encryptCredentials(creds: Record<string, unknown>): string {
  const iv = randomBytes(16);
  const plaintext = Buffer.from(JSON.stringify(creds), "utf8");
  const cipher = createCipheriv("aes-128-cbc", encryptionKey, iv);
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const payload = Buffer.concat([
    Buffer.from([0x80]),
    fernetTimestamp(),
    iv,
    ciphertext
  ]);
  const signature = createHmac("sha256", signingKey).update(payload).digest();
  return Buffer.concat([payload, signature]).toString("base64url");
}

export function decryptCredentials(encrypted: string): Record<string, unknown> {
  const token = Buffer.from(encrypted, "base64url");
  const payload = token.subarray(0, token.length - 32);
  const signature = token.subarray(token.length - 32);
  const expected = createHmac("sha256", signingKey).update(payload).digest();
  if (!signature.equals(expected)) {
    throw new Error("Invalid credential signature");
  }

  const iv = payload.subarray(9, 25);
  const ciphertext = payload.subarray(25);
  const decipher = createDecipheriv("aes-128-cbc", encryptionKey, iv);
  const decrypted = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final()
  ]);

  return JSON.parse(decrypted.toString("utf8")) as Record<
    string,
    unknown
  >;
}

export async function hashPassword(password: string): Promise<string> {
  return bcrypt.hash(password, 10);
}

export async function verifyPassword(
  password: string,
  hashedPassword: string
): Promise<boolean> {
  return bcrypt.compare(password, hashedPassword);
}

export async function createAccessToken(data: Record<string, string>) {
  const now = Math.floor(Date.now() / 1000);
  return new SignJWT({ ...data, type: "access" })
    .setProtectedHeader({ alg: "HS256" })
    .setIssuedAt(now)
    .setExpirationTime(now + config.accessTokenExpireMinutes * 60)
    .sign(jwtKey);
}

export function createRefreshToken(): string {
  return randomBytes(48).toString("hex");
}

export async function decodeAccessToken(token: string) {
  try {
    const { payload } = await jwtVerify(token, jwtKey, {
      algorithms: ["HS256"]
    });
    if (payload.type !== "access") {
      return null;
    }
    return payload;
  } catch {
    return null;
  }
}

export function sha256(content: string): string {
  return createHash("sha256").update(content).digest("hex");
}

File: backend/app/utils/security.py
from __future__ import annotations

import hashlib
import hmac
import json
import os
import struct
import time
from typing import Any

import bcrypt as _bcrypt
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import padding as crypto_padding
from jose import JWTError, jwt

from app.core.config import settings


def _get_fernet_key_bytes() -> bytes:
    key = settings.encryption_key
    try:
        import base64
        decoded = base64.urlsafe_b64decode(key + "==")
        if len(decoded) == 32:
            return decoded
    except Exception:
        pass
    return hashlib.sha256(key.encode("utf-8")).digest()


def _get_key_material() -> tuple[bytes, bytes]:
    key_bytes = _get_fernet_key_bytes()
    signing_key = key_bytes[:16]
    encryption_key = key_bytes[16:32]
    return signing_key, encryption_key


def encrypt_credentials(creds: dict[str, Any]) -> str:
    import base64

    signing_key, encryption_key = _get_key_material()
    iv = os.urandom(16)
    plaintext = json.dumps(creds).encode("utf-8")

    padder = crypto_padding.PKCS7(128).padder()
    padded = padder.update(plaintext) + padder.finalize()

    cipher = Cipher(algorithms.AES(encryption_key), modes.CBC(iv))
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(padded) + encryptor.finalize()

    ts = int(time.time())
    timestamp = struct.pack(">Q", ts)
    payload = bytes([0x80]) + timestamp + iv + ciphertext

    signature = hmac.new(signing_key, payload, hashlib.sha256).digest()
    token = payload + signature
    return base64.urlsafe_b64encode(token).rstrip(b"=").decode("ascii")


def decrypt_credentials(encrypted: str) -> dict[str, Any]:
    import base64

    signing_key, encryption_key = _get_key_material()

    padding = (4 - len(encrypted) % 4) % 4
    token = base64.urlsafe_b64decode(encrypted + "=" * padding)

    payload = token[:-32]
    signature = token[-32:]
    expected = hmac.new(signing_key, payload, hashlib.sha256).digest()
    if not hmac.compare_digest(signature, expected):
        raise ValueError("Invalid credential signature")

    iv = payload[9:25]
    ciphertext = payload[25:]

    cipher = Cipher(algorithms.AES(encryption_key), modes.CBC(iv))
    decryptor = cipher.decryptor()
    padded = decryptor.update(ciphertext) + decryptor.finalize()

    unpadder = crypto_padding.PKCS7(128).unpadder()
    decrypted = unpadder.update(padded) + unpadder.finalize()
    return json.loads(decrypted.decode("utf-8"))


def hash_password(password: str) -> str:
    return _bcrypt.hashpw(password.encode("utf-8"), _bcrypt.gensalt(rounds=10)).decode("utf-8")


def verify_password(password: str, hashed: str) -> bool:
    return _bcrypt.checkpw(password.encode("utf-8"), hashed.encode("utf-8"))


def create_access_token(data: dict[str, str]) -> str:
    now = int(time.time())
    payload = {
        **data,
        "type": "access",
        "iat": now,
        "exp": now + settings.access_token_expire_minutes * 60,
    }
    return jwt.encode(payload, settings.app_secret, algorithm="HS256")


def create_refresh_token() -> str:
    return os.urandom(48).hex()


def decode_access_token(token: str) -> dict[str, Any] | None:
    try:
        payload = jwt.decode(token, settings.app_secret, algorithms=["HS256"])
        if payload.get("type") != "access":
            return None
        return payload
    except JWTError:
        return None


def sha256(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()
